web/src/main.js

```diff
@@ -927,16 +927,20 @@ function dataLoaded(err, clientVersion, argCallerLocation, state, argSolar, argW
   // Is there a new version?
   // d3.select('#new-version')
   //   .classed('active', (clientVersion !== bundleHash && !store.getState().application.isLocalhost && !isCordova));
-
   // TODO: Code is duplicated
-  currentMoment = (getState().application.customDate && moment(getState().application.customDate) || moment(state.datetime));
-  d3.selectAll('.current-datetime').text(currentMoment.format('LL LT'));
-  d3.selectAll('.current-datetime-from-now')
-    .text(currentMoment.fromNow())
-    .style('color', 'darkred')
-    .transition()
-    .duration(800)
-    .style('color', undefined);
+  const setLastUpdated = () => {
```

Comment:
I think you're inside the method `dataLoaded`, which is called every time the frontend has new data.
This will create several `setInterval`.
You should probably move this outside of the method, and just call `setLastUpdated` from inside.

Reply:
Ah apologies!

Ok if I move the function outside of `dataLoaded` there is a dependency against the state object due to the following line
```javascript
  currentMoment = (getState().application.customDate && moment(getState().application.customDate) || moment(state.datetime));
````
This fine when calling `setLastUpdated` from within `dataLoaded`, but if the function is called outside `dataLoaded` what can be used to pass the `state` parameter ?

Reply:
I think the best is to use the new redux system we're putting into place (but didn't finish yet)

In a nutshell, you should
- add a new `case` in the reducer (in src/reducers/main.js) to handle a new `GRID_DATA` event, that updates the current application state, setting `data.grid` to the supplied state object (note: the application state and the grid state are two different objects)
- in `dataLoaded`, submit the grid state object to the application state by calling
```js
dispatch({
  payload: state,
  type: 'GRID_DATA',
});
```
- you will then be able to get the current date from `setLastUpdated` by calling `getState().grid.datetime`
- you should subscribe to datetime changes in order to call `setLastUpdated`. You can do that in the main.js using the `observe` method (look at existing implementations to see an example)
- set the setInterval method outside of the `dataLoaded`